File: Ease the StockBroker/ease-stockbroker.php
<?php
/**
 * Created by PhpStorm.
 * User: kovalevcon
 * Date: 04.09.18
 * Time: 11:44
 */

/**
 * Ease the StockBroker function
 *
 * @param string $list
 * @return string
 */
function balanceStatements(string $list): string
{
    if (!strlen($list)) {
        return 'Buy: 0 Sell: 0';
    }

    $buySum = $sellSum = 0;
    $badlyFormed = [];

    $items = explode(", ", $list);
    foreach ($items as $item) {
        $order = explode(' ', $item);
        if (isCorrectOrder($order)) {
            $total = $order[1] * (float)$order[2];
            $order[3] == 'B' ? $buySum += $total : $sellSum += $total;
        } else {
            $badlyFormed[] = "{$item} ";
        }
    }

    $result = sprintf("Buy: %d Sell: %d", round($buySum), round($sellSum));
    $countBadly = count($badlyFormed);
    if ($countBadly) {
        $result .= sprintf("; Badly formed %d: %s;", $countBadly, implode(";", $badlyFormed));
    }

    return $result;
}

function isCorrectOrder(array $order): bool
{
    return count($order) == 4 && is_string($order[0]) && preg_match('/^[\d]+$/', $order[1])
        && preg_match('/^[\d]*\.[\d]+$/', $order[2])
        && in_array($order[3], ['B','S']);
}

echo "<pre>balanceStatements('GOOG 300 542.0 B, AAPL 50 145.0 B, CSCO 250.0 29 B, GOOG 200 580.0 S')\r\n";
echo balanceStatements('GOOG 300 542.0 B, AAPL 50 145.0 B, CSCO 250.0 29 B, GOOG 200 580.0 S'); // Buy: 169850 Sell: 116000; Badly formed 1: CSCO 250.0 29 B ;
echo "</pre>";

echo "<pre>balanceStatements('')\r\n";
echo balanceStatements(''); // Buy: 0 Sell: 0
echo "</pre>";

File: Ease the StockBroker/ease-stockbroker.py
import re


def balance_statement(lst):
    buy_sum = 0.0
    sell_sum = 0.0
    badly = []

    if len(lst):
        for item in lst.split(', '):
            if re.match('^[a-zA-Z0-9.]+\s\d+\s\d*\.\d+\s[BS]$', item):
                order = item.split(' ')
                total = int(order[1]) * float(order[2])
                if order[3] == 'B':
                    buy_sum += total
                else:
                    sell_sum += total
            else:
                badly.append("%s " % item)

    result = "Buy: %d Sell: %d" % (round(buy_sum), round(sell_sum))
    if len(badly) > 0:
        result += "; Badly formed %d: %s;" % (len(badly), ";".join(badly))

    return result


# Buy: 29499 Sell: 0
print(balance_statement('ZNGA 1300 2.66 B, CLH15.NYM 50 56.32 B, OWW 1000 11.623 B, OGG 20 580.1 B'))
